From 2bf5843647b5ccd6549175bc3a40a48b73f9013c Mon Sep 17 00:00:00 2001
From: CL Chin-Long Wang <cl634@atctrx.andestech.com>
Date: Fri, 19 May 2023 14:51:53 +0800
Subject: [PATCH 040/191] soc: andes: atcdmac300: fix channel pending when
 burn-in the dmac driver

1. Fixed the issue "Cannot change direction while the channel has pending"
   while burn-in the dmac driver.
2. Remove "raw_spin_lock / raw_spin_unlock" from legacy_dma_irq_router
3. Add a handle procedure to handle the status error
   while "DMAD_FLAGS_RING_MODE".
4. It is suggested in the Gitea code review:
	https://gitea.andestech.com/RD-SW/linux/pulls/17#issuecomment-73741

Signed-off-by: CL Wang <cl634@andestech.com>
---
 drivers/mmc/host/ftsdc010.c    |  22 +++--
 drivers/soc/andes/atcdmac300.c | 176 ++++++++++++++++++++++-----------
 drivers/soc/andes/dmad_intc.c  |   8 +-
 include/soc/andes/atcdmac300.h |   5 +-
 4 files changed, 135 insertions(+), 76 deletions(-)

diff --git a/drivers/mmc/host/ftsdc010.c b/drivers/mmc/host/ftsdc010.c
index ed43d596d61a..220d041e1b63 100644
--- a/drivers/mmc/host/ftsdc010.c
+++ b/drivers/mmc/host/ftsdc010.c
@@ -401,12 +401,13 @@ static void do_pio_write(struct ftsdc_host *host)
 	host->complete_what = COMPLETION_FINALIZE;
 }
 
-static void do_dma_access(struct ftsdc_host *host)
+static void do_dma_access(struct ftsdc_host *host, struct mmc_data *data)
 {
 	int res;
 	unsigned long timeout;
 	dmad_chreq *req = host->dma_req;
 	dmad_drb *drb = 0;
+	int rw = (data->flags & MMC_DATA_WRITE) ? 1 : 0;
 
 	while (host->buf_sgptr < host->mrq->data->sg_len) {
 
@@ -431,6 +432,15 @@ static void do_dma_access(struct ftsdc_host *host)
 		drb->addr1 = (dma_addr_t) (unsigned long)host->buf_ptr;
 		drb->req_cycle = dmad_bytes_to_cycles(req, host->buf_bytes);
 		drb->sync = &host->dma_complete;
+
+		res = dmad_config_channel_dir(host->dma_req,
+					      rw ? DMAD_DIR_A1_TO_A0 : DMAD_DIR_A0_TO_A1,
+					      drb);
+		if (res != 0) {
+			host->mrq->data->error = -ENODEV;
+			goto err;
+		}
+
 		timeout = SDC_TIMEOUT_BASE * ((host->buf_bytes + 511) >> 9);
 		res = dmad_submit_request(req, drb, 1);
 		if (res != 0) {
@@ -464,7 +474,7 @@ static void ftsdc_work(struct work_struct *work)
 
 	ftsdc_enable_irq(host, false);
 	if (host->dodma) {
-		do_dma_access(host);
+		do_dma_access(host, data);
 		dma_unmap_sg(mmc_dev(host->mmc), data->sg, data->sg_len,
 			     rw ? DMA_TO_DEVICE : DMA_FROM_DEVICE);
 	} else {
@@ -722,20 +732,12 @@ static int ftsdc_prepare_buffer(struct ftsdc_host *host, struct mmc_data *data)
 	host->buf_active = rw ? XFER_WRITE : XFER_READ;
 	if (host->dodma) {
 		u32 dma_len;
-		u32 drb_size;
 
 		dma_len = dma_map_sg(mmc_dev(host->mmc), data->sg, data->sg_len,
 				     rw ? DMA_TO_DEVICE : DMA_FROM_DEVICE);
 		if (dma_len == 0)
 			return -ENOMEM;
 
-		dmad_config_channel_dir(host->dma_req,
-					rw ? DMAD_DIR_A1_TO_A0 :
-					DMAD_DIR_A0_TO_A1);
-		drb_size = dmad_max_size_per_drb(host->dma_req);
-		if (drb_size < (data->blksz & data->blocks))
-			return -ENODEV;
-
 		host->dma_finish = false;
 	}
 	return 0;
diff --git a/drivers/soc/andes/atcdmac300.c b/drivers/soc/andes/atcdmac300.c
index 1089938e3943..1c704099156e 100644
--- a/drivers/soc/andes/atcdmac300.c
+++ b/drivers/soc/andes/atcdmac300.c
@@ -1,7 +1,6 @@
 // SPDX-License-Identifier: GPL-2.0
 /*
  * Copyright (C) 2018 Andes Technology Corporation
- *
  */
 
 #include <linux/module.h>
@@ -41,12 +40,12 @@ static inline void REG_WRITE(addr_t d, unsigned long r)
 
 #define DMAD_DRB_POOL_SIZE 32	/* 128 */
 
-static inline addr_t din(unsigned long r)
+static inline u32 din(unsigned long r)
 {
 	return REG_READ(r);
 }
 
-static inline void dout(addr_t d, unsigned long r)
+static inline void dout(u32 d, unsigned long r)
 {
 	REG_WRITE(d, r);
 }
@@ -86,6 +85,8 @@ typedef struct dmad_drq {
 	unsigned long enable_port;	/* enable register */
 	unsigned long src_port;	/* source address register */
 	unsigned long dst_port;	/* dest address register */
+	unsigned long src_port_h;	/* source address register (high part) */
+	unsigned long dst_port_h;	/* dest address register (high part) */
 	unsigned long cyc_port;	/* size(cycle) register */
 
 	u32 flags;		/* enum DMAD_CHREQ_FLAGS */
@@ -127,10 +128,10 @@ typedef struct dmad_drq {
 	/* ring_size - period_size * periods */
 	dma_addr_t remnant_size;
 
-	dma_addr_t sw_ptr;	/* sw pointer */
+	u64 sw_ptr;		/* sw pointer */
 	int sw_p_idx;		/* current ring_ptr */
 	dma_addr_t sw_p_off;	/* offset to period base */
-
+	int channel;
 } dmad_drq;
 
 static inline void dmad_enable_channel(dmad_drq *drq)
@@ -148,12 +149,6 @@ static inline addr_t dmad_is_channel_enabled(dmad_drq *drq)
 	return (addr_t) getbl(CHEN, drq->enable_port);
 }
 
-/* system irq number (per channel, ahb) */
-static const unsigned int ahb_irqs[DMAD_AHB_MAX_CHANNELS] = {
-	DMA_IRQ0, DMA_IRQ1, DMA_IRQ2, DMA_IRQ3,
-	DMA_IRQ4, DMA_IRQ5, DMA_IRQ6, DMA_IRQ7,
-};
-
 /* Driver data structure, one instance per system */
 typedef struct DMAD_DATA_STRUCT {
 	/* Driver data initialization flag */
@@ -514,7 +509,7 @@ static irqreturn_t dmad_ahb_isr(int irq, void *dev_id)
 	}
 
 	/* Process DRBs according to interrupt reason */
-	if (tc_int) {
+	if (tc_int || ((drq->flags & DMAD_FLAGS_RING_MODE) && err_int)) {
 		dmad_dbg("dma finish\n");
 		dmad_dbg("finish drb(%d 0x%08x) addr0(0x%08llx) addr1 (0x%08llx) size(0x%08llx)\n",
 			drb->node, (u32) drb, drb->src_addr, drb->dst_addr,
@@ -544,20 +539,33 @@ static irqreturn_t dmad_ahb_isr(int irq, void *dev_id)
 			// Kick-off DMA for next DRB
 			// - Source and destination address
 			if (drq->flags & DMAD_DRQ_DIR_A1_TO_A0) {
-				dout(drb_iter->addr1,
-				     (unsigned long)drq->src_port);
-				dout(drb_iter->addr0,
-				     (unsigned long)drq->dst_port);
+				dout(drb_iter->addr1, (unsigned long)drq->src_port);
+				dout(drb_iter->addr0, (unsigned long)drq->dst_port);
+
+#ifdef CONFIG_ARCH_DMA_ADDR_T_64BIT
+				dout((drb_iter->addr1 >> 32), (unsigned long)drq->src_port_h);
+				dout((drb_iter->addr0 >> 32), (unsigned long)drq->dst_port_h);
+#else
+				dout(0, (unsigned long)drq->src_port_h);
+				dout(0, (unsigned long)drq->dst_port_h);
+#endif
 			} else {
-				dout(drb_iter->addr0,
-				     (unsigned long)drq->src_port);
-				dout(drb_iter->addr1,
-				     (unsigned long)drq->dst_port);
+				dout(drb_iter->addr0, (unsigned long)drq->src_port);
+				dout(drb_iter->addr1, (unsigned long)drq->dst_port);
+
+#ifdef CONFIG_ARCH_DMA_ADDR_T_64BIT
+				dout((drb_iter->addr0 >> 32), (unsigned long)drq->src_port_h);
+				dout((drb_iter->addr1 >> 32), (unsigned long)drq->dst_port_h);
+#else
+				dout(0, (unsigned long)drq->src_port_h);
+				dout(0, (unsigned long)drq->dst_port_h);
+#endif
 			}
 
 			/* - Transfer size (in units of source width) */
 			dout(drb_iter->req_cycle, (unsigned long)drq->cyc_port);
 
+			dout(drb->ch_control, (unsigned long)drq->enable_port);
 			/* Kick off next request */
 			dmad_enable_channel(drq);
 
@@ -633,7 +641,7 @@ static void dmad_ahb_config_dir(dmad_chreq *ch_req,
 	dmad_dbg("%s() channel_cmds(0x%08lx)\n", __func__, channel_cmds[0]);
 	channel_cmds[0] &= ~(u32) (SRCWIDTH_MASK | SRCADDRCTRL_MASK |
 				   DSTWIDTH_MASK | DSTADDRCTRL_MASK | SRC_HS |
-				   DST_HS | SRCREQSEL_MASK | DSTREQSEL_MASK);
+				   DST_HS | SRCREQSEL_MASK | DSTREQSEL_MASK | (1 << CHEN));
 
 	if (ahb_req->tx_dir == 0) {
 		dmad_dbg("%s() addr0 --> addr1\n", __func__);
@@ -693,8 +701,7 @@ static int dmad_ahb_init(dmad_chreq *ch_req)
 	err = request_irq(pdata->irqs[channel + 1], dmad_ahb_isr, 0, "AHB_DMA",
 			(void *)(unsigned long)(channel + 1));
 	if (unlikely(err != 0)) {
-		dmad_err("unable to request IRQ %d for AHB DMA (error %d)\n",
-			 ahb_irqs[channel], err);
+		dmad_err("unable to request IRQ for channel:%d err:%d\n", channel, err);
 		return err;
 	}
 	spin_lock_irqsave(&dmad.drq_pool_lock, lock_flags);
@@ -725,11 +732,15 @@ static int dmad_ahb_init(dmad_chreq *ch_req)
 
 	/* SRCADR and DESADR */
 	dout(0, (unsigned long)drq->src_port);
+	dout(0, (unsigned long)drq->src_port_h);
+
 	dout(0, (unsigned long)drq->dst_port);
+	dout(0, (unsigned long)drq->dst_port_h);
 	/* CYC (transfer size) */
 	dout(0, (unsigned long)drq->cyc_port);
 	/* LLP */
 	dout(0, (unsigned long)channel_base + CH_LLP_LOW_OFF);
+	dout(0, (unsigned long)channel_base + CH_LLP_HIGH_OFF);
 
 	/* TOT_SIZE - not now */
 	spin_unlock_irqrestore(&dmad.drq_pool_lock, lock_flags);
@@ -900,12 +911,15 @@ int dmad_channel_alloc(dmad_chreq *ch_req)
 	/* Record the channel's queue handle in client's struct */
 	ch_req->drq = drq_iter;
 
+	drq_iter->channel = i;
 	if (ch_req->controller == DMAD_DMAC_AHB_CORE) {
 		drq_iter->channel_base = (unsigned long)DMAC_BASE_CH(i);
 		drq_iter->enable_port = (unsigned long)CH_CTL(i);
 		drq_iter->src_port = (unsigned long)CH_SRC_L(i);
 		drq_iter->dst_port = (unsigned long)CH_DST_L(i);
 		drq_iter->cyc_port = (unsigned long)CH_SIZE(i);
+		drq_iter->src_port_h = (unsigned long)CH_SRC_H(i);
+		drq_iter->dst_port_h = (unsigned long)CH_DST_H(i);
 	}
 	/* drb-0 is an invalid node - for node validation */
 	drb_iter = &drq_iter->drb_pool[0];
@@ -943,6 +957,13 @@ int dmad_channel_alloc(dmad_chreq *ch_req)
 			goto _err_exit;
 		}
 
+		drb_iter = &drq_iter->drb_pool[0];
+		for (i = 0; i < DMAD_DRB_POOL_SIZE; ++i, ++drb_iter) {
+			err = dmad_config_channel_dir(ch_req, ch_req->ahb_req.tx_dir, drb_iter);
+			if (err != 0)
+				goto _err_exit;
+		}
+
 		drq_iter->ring_size = ch_req->ring_size;
 		drq_iter->period_size = ch_req->period_size;
 		drq_iter->remnant_size = (dma_addr_t) remnant;
@@ -1110,52 +1131,25 @@ EXPORT_SYMBOL_GPL(dmad_channel_enable);
  * that bi-direction mode and ring mode are mutual-exclusive from user's
  * perspective.
  */
-int dmad_config_channel_dir(dmad_chreq *ch_req, u8 dir)
+int dmad_config_channel_dir(dmad_chreq *ch_req, u8 dir, dmad_drb *drb)
 {
 	dmad_drq *drq;
 	addr_t channel_cmds[1];
-	unsigned long lock_flags;
-	u8 cur_dir;
 
 	if (unlikely(ch_req == NULL))
 		return -EFAULT;
 
 	drq = (dmad_drq *) ch_req->drq;
-
 	if (unlikely(drq == NULL))
 		return -EBADR;
 
-	if (unlikely(!(ch_req->flags & DMAD_FLAGS_BIDIRECTION))) {
-		dmad_err("%s() Channel is not configured as	bidirectional!\n",
-			__func__);
-		return -EFAULT;
-	}
-
-	cur_dir = drq->flags & DMAD_DRQ_DIR_MASK;
-	if (dir == cur_dir) {
-		dmad_dbg("%s() cur_dir(%d) == dir(%d) skip reprogramming hw.\n",
-			__func__, cur_dir, dir);
-		return 0;
-	}
-
-	spin_lock_irqsave(&drq->drb_pool_lock, lock_flags);
-
-	if (unlikely((drq->sbt_head != 0) /*||dmad_is_channel_enabled(drq) */)) {
-		spin_unlock_irqrestore(&drq->drb_pool_lock, lock_flags);
-		dmad_err("%s() Cannot change direction while the channel has pending requests!\n",
-			__func__);
-		return -EFAULT;
-	}
-
 	if (ch_req->controller == DMAD_DMAC_AHB_CORE) {
 		channel_cmds[0] = din((unsigned long)drq->enable_port);
 		ch_req->ahb_req.tx_dir = dir;
 		dmad_ahb_config_dir(ch_req, (unsigned long *)channel_cmds);
-		dout(channel_cmds[0], (unsigned long)drq->enable_port);
+		drb->ch_control = channel_cmds[0];
 	}
 
-	spin_unlock_irqrestore(&drq->drb_pool_lock, lock_flags);
-
 	return 0;
 }
 EXPORT_SYMBOL_GPL(dmad_config_channel_dir);
@@ -1520,13 +1514,30 @@ int dmad_submit_request(dmad_chreq *ch_req, dmad_drb *drb, u8 keep_fired)
 		if (drq->flags & DMAD_DRQ_DIR_A1_TO_A0) {
 			dout(drb->addr1, (unsigned long)drq->src_port);
 			dout(drb->addr0, (unsigned long)drq->dst_port);
+
+#ifdef CONFIG_ARCH_DMA_ADDR_T_64BIT
+			dout((drb->addr1 >> 32), (unsigned long)drq->src_port_h);
+			dout((drb->addr0 >> 32), (unsigned long)drq->dst_port_h);
+#else
+			dout(0, (unsigned long)drq->src_port_h);
+			dout(0, (unsigned long)drq->dst_port_h);
+#endif
 		} else {
 			dout(drb->addr0, (unsigned long)drq->src_port);
 			dout(drb->addr1, (unsigned long)drq->dst_port);
+
+#ifdef CONFIG_ARCH_DMA_ADDR_T_64BIT
+			dout((drb->addr0 >> 32), (unsigned long)drq->src_port_h);
+			dout((drb->addr1 >> 32), (unsigned long)drq->dst_port_h);
+#else
+			dout(0, (unsigned long)drq->src_port_h);
+			dout(0, (unsigned long)drq->dst_port_h);
+#endif
 		}
 
 		/* Transfer size (in units of source width) */
 		dout(drb->req_cycle, (unsigned long)drq->cyc_port);
+		dout(drb->ch_control, (unsigned long)drq->enable_port);
 
 		/* Enable DMA channel (Kick off transmission when client
 		 * enable it's transfer state)
@@ -1648,17 +1659,32 @@ static inline int dmad_kickoff_requests_internal(dmad_drq *drq)
 		drb->req_cycle, drb->state);
 
 	if (drb->state == DMAD_DRB_STATE_SUBMITTED) {
-		/* Transfer size (in units of source width) */
-		dout(drb->req_cycle, (unsigned long)drq->cyc_port);
 
 		/* Source and destination address */
 		if (drq->flags & DMAD_DRQ_DIR_A1_TO_A0) {
 			dout(drb->addr1, (unsigned long)drq->src_port);
 			dout(drb->addr0, (unsigned long)drq->dst_port);
+#ifdef CONFIG_ARCH_DMA_ADDR_T_64BIT
+			dout((drb->addr1 >> 32), (unsigned long)drq->src_port_h);
+			dout((drb->addr0 >> 32), (unsigned long)drq->dst_port_h);
+#else
+			dout(0, (unsigned long)drq->src_port_h);
+			dout(0, (unsigned long)drq->dst_port_h);
+#endif
 		} else {
 			dout(drb->addr0, (unsigned long)drq->src_port);
 			dout(drb->addr1, (unsigned long)drq->dst_port);
+#ifdef CONFIG_ARCH_DMA_ADDR_T_64BIT
+			dout((drb->addr0 >> 32), (unsigned long)drq->src_port_h);
+			dout((drb->addr1 >> 32), (unsigned long)drq->dst_port_h);
+#else
+			dout(0, (unsigned long)drq->src_port_h);
+			dout(0, (unsigned long)drq->dst_port_h);
+#endif
 		}
+		/* Transfer size (in units of source width) */
+		dout(drb->req_cycle, (unsigned long)drq->cyc_port);
+		dout(drb->ch_control, (unsigned long)drq->enable_port);
 
 		drb->state = DMAD_DRB_STATE_EXECUTED;
 	}
@@ -1724,18 +1750,35 @@ int dmad_kickoff_requests(dmad_chreq *ch_req)
 		goto _safe_exit;
 	}
 
-	/* Transfer size (in units of source width) */
-	dout(req_cycle, (unsigned long)drq->cyc_port);
-
 	/* Source and destination address */
 	if (drq->flags & DMAD_DRQ_DIR_A1_TO_A0) {
 		dout(drb->addr1, (unsigned long)drq->src_port);
 		dout(drb->addr0, (unsigned long)drq->dst_port);
+
+#ifdef CONFIG_ARCH_DMA_ADDR_T_64BIT
+		dout((drb->addr1 >> 32), (unsigned long)drq->src_port_h);
+		dout((drb->addr0 >> 32), (unsigned long)drq->dst_port_h);
+#else
+		dout(0, (unsigned long)drq->src_port_h);
+		dout(0, (unsigned long)drq->dst_port_h);
+#endif
+
 	} else {
 		dout(drb->addr0, (unsigned long)drq->src_port);
 		dout(drb->addr1, (unsigned long)drq->dst_port);
+
+#ifdef CONFIG_ARCH_DMA_ADDR_T_64BIT
+		dout((drb->addr0 >> 32), (unsigned long)drq->src_port_h);
+		dout((drb->addr1 >> 32), (unsigned long)drq->dst_port_h);
+#else
+		dout(0, (unsigned long)drq->src_port_h);
+		dout(0, (unsigned long)drq->dst_port_h);
+#endif
 	}
+	/* Transfer size (in units of source width) */
+	dout(req_cycle, (unsigned long)drq->cyc_port);
 
+	dout(drb->ch_control, (unsigned long)drq->enable_port);
 	drb->state = DMAD_DRB_STATE_EXECUTED;
 
 	/* Enable DMA channel */
@@ -1850,7 +1893,8 @@ int dmad_update_ring_sw_ptr(dmad_chreq *ch_req, dma_addr_t sw_ptr,
 	dmad_drq *drq;
 	unsigned long lock_flags;
 	dma_addr_t hw_off = 0, ring_ptr;
-	dma_addr_t sw_p_off, ring_p_off, period_bytes;
+	u64 sw_p_off;
+	dma_addr_t ring_p_off, period_bytes;
 	dma_addr_t remnant_size;
 	int period_size;
 	int sw_p_idx, ring_p_idx, period, periods;
@@ -1912,6 +1956,11 @@ int dmad_update_ring_sw_ptr(dmad_chreq *ch_req, dma_addr_t sw_ptr,
 		    drq->ring_base;
 		drb->addr1 = drq->dev_addr;
 		drb->req_cycle = 0;	// redundent, though, no harm to performance
+		if (dmad_config_channel_dir(ch_req, ch_req->ahb_req.tx_dir, drb) != 0) {
+			dmad_err("%s() dmad_config_channel_dir failed!\n", __func__);
+			spin_unlock_irqrestore(&drq->drb_pool_lock, lock_flags);
+			return -ENOSPC;
+		}
 
 		dmad_dbg("init_drb(%d 0x%08x) addr0(0x%08llx) addr1(0x%08llx) size(0x%08llx) state(%d)\n",
 			(u32) drb->node, (u32) drb, drb->src_addr,
@@ -2077,6 +2126,12 @@ int dmad_update_ring_sw_ptr(dmad_chreq *ch_req, dma_addr_t sw_ptr,
 			else
 				drb->req_cycle = period_size;
 
+			if (dmad_config_channel_dir(ch_req, ch_req->ahb_req.tx_dir, drb) != 0) {
+				dmad_err("%s() dmad_config_channel_dir failed!\n", __func__);
+				spin_unlock_irqrestore(&drq->drb_pool_lock, lock_flags);
+				return -ENOSPC;
+			}
+
 			dmad_dbg("inbtw_drb(%d 0x%08x) addr0(0x%08llx)addr1 (0x%08llx) size(0x%08llx) state(%d)\n",
 				 (u32) drb->node, (u32) drb, drb->addr0,
 				 drb->addr1, drb->req_cycle, drb->state);
@@ -2094,6 +2149,11 @@ int dmad_update_ring_sw_ptr(dmad_chreq *ch_req, dma_addr_t sw_ptr,
 		    drq->ring_base;
 		drb->addr1 = drq->dev_addr;
 		drb->req_cycle = sw_p_off;
+		if (dmad_config_channel_dir(ch_req, ch_req->ahb_req.tx_dir, drb) != 0) {
+			dmad_err("%s() dmad_config_channel_dir failed!\n", __func__);
+			spin_unlock_irqrestore(&drq->drb_pool_lock, lock_flags);
+			return -ENOSPC;
+		}
 
 		dmad_dbg("swptr_drb(%d 0x%08x) addr0(0x%08llx) addr1(0x%08llx) size(0x%08llx) state(%d)\n",
 			 (u32) drb->node, (u32) drb, drb->addr0, drb->addr1,
diff --git a/drivers/soc/andes/dmad_intc.c b/drivers/soc/andes/dmad_intc.c
index 0e9d0f5499e5..46c2fdadc9b4 100644
--- a/drivers/soc/andes/dmad_intc.c
+++ b/drivers/soc/andes/dmad_intc.c
@@ -10,25 +10,21 @@
 
 #ifdef CONFIG_PLATFORM_AHBDMA
 
-void AHBDMA_irq_rounter(struct irq_desc *desc)
+static void legacy_dma_irq_router(struct irq_desc *desc)
 {
 	int ahb_irq;
 	struct irq_desc *ahb_desc;
 
-	raw_spin_lock(&desc->lock);
 	ahb_irq = dmad_probe_irq_source_ahb();
 
 	if (ahb_irq >= 0) {
 		ahb_irq = get_irq(ahb_irq);
 		ahb_desc = irq_to_desc(ahb_irq);
 		ahb_desc->irq_data.irq = ahb_irq;
-		raw_spin_unlock(&desc->lock);
 		ahb_desc->handle_irq(ahb_desc);
-		raw_spin_lock(&desc->lock);
 	}
 	desc->irq_data.chip->irq_unmask(&desc->irq_data);
 	desc->irq_data.chip->irq_eoi(&desc->irq_data);
-	raw_spin_unlock(&desc->lock);
 }
 
 int intc_ftdmac020_init_irq(int irq)
@@ -41,7 +37,7 @@ int intc_ftdmac020_init_irq(int irq)
 		ret = irq_set_chip(i, &dummy_irq_chip);
 		irq_set_handler(i, handle_simple_irq);
 	}
-	irq_set_chained_handler(irq, AHBDMA_irq_rounter);
+	irq_set_chained_handler(irq, legacy_dma_irq_router);
 	return 0;
 }
 #endif /* CONFIG_PLATFORM_AHBDMA */
diff --git a/include/soc/andes/atcdmac300.h b/include/soc/andes/atcdmac300.h
index cbb81b4e6062..39d501f6701c 100644
--- a/include/soc/andes/atcdmac300.h
+++ b/include/soc/andes/atcdmac300.h
@@ -423,6 +423,7 @@ typedef struct dmad_drb {
 	 */
 	dma_addr_t req_cycle;
 
+	u32 ch_control;
 	/* (in)  if non-null, this sync object will be signaled upon dma
 	 * completion (for blocked-waiting dma completion)
 	 */
@@ -476,7 +477,7 @@ extern dma_addr_t dmad_probe_hw_ptr_dst(dmad_chreq *ch_req);
 /*****************************************************************************
  * routines only valid in discrete (non-ring) mode
  */
-extern int dmad_config_channel_dir(dmad_chreq *ch_req, u8 dir);
+extern int dmad_config_channel_dir(dmad_chreq *ch_req, u8 dir, dmad_drb *drb);
 extern int dmad_alloc_drb(dmad_chreq *ch_req, dmad_drb **drb);
 extern int dmad_free_drb(dmad_chreq *ch_req, dmad_drb *drb);
 extern int dmad_submit_request(dmad_chreq *ch_req,
@@ -514,7 +515,7 @@ static inline dma_addr_t dmad_probe_hw_ptr_src(dmad_chreq *ch_req)
 	{ return (dma_addr_t)NULL; }
 static inline dma_addr_t dmad_probe_hw_ptr_dst(dmad_chreq *ch_req)
 	{ return (dma_addr_t)NULL; }
-static inline int dmad_config_channel_dir(dmad_chreq *ch_req, u8 dir)
+static inline int dmad_config_channel_dir(dmad_chreq *ch_req, u8 dir, dmad_drb *drb)
 	{ return -EFAULT; }
 static inline int dmad_alloc_drb(dmad_chreq *ch_req, dmad_drb **drb)
 	{ return -EFAULT; }
-- 
2.34.1

